src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(LSPDocumentSynchronizer))]
+    internal class DefaultLSPDocumentSynchronizer : LSPDocumentSynchronizer
+    {
+        // Internal for testing
+        internal TimeSpan _synchronizationTimeout = TimeSpan.FromSeconds(2);
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly ConcurrentDictionary<Uri, DocumentSynchronizingContext> _synchronizingContexts;
```

Comment:
Does using Uri as key here work well for documents from network shares?

Reply:
Why wouldn't it? I ask because the Uri infrastructure is what the LSP pieces of VS are built-on

Reply:
I'm not sure how it is works here but when the language server receives a CompletionRequest for example, if you inspect the URI, the Uri.Host will be all lowercase where as in VSCode if you do `vscode.workspace.findFiles`, the Uris it returns will have `Uri.Authority`(Host equivalent) in whatever-case. So I had to normalize it in VSCode. I was wondering if you will run into similar problems here.

Reply:
Hmm, interesting. I'm not entirely sure. If we do need to do that though it'll be done at the LSP document factory layers where the Uris are constructed